### PR TITLE
[REVIEW] v1.4.0 Code Review

### PR DESCRIPTION
**Pull request by bryanculver**
A LOT of changes have come in 1.4.0 with models, serializers, filters, etc.

Opening this PR as draft to provide a way to do collaborative code review of `next` without having to have huge CI churn that exists with merging into `next` and `develop`.

This is NOT meant to be the final PR, any changes we incorporate here we can merge into `next` on a semi-regular basis.

CI is disabled to eliminate churn on every new merge to `next`.

Please leave comments as you see fit. If you would like to make a change to get folded into `next`, either open a separate PR into `next` or to this branch.

image

### TODO

- [x] Sensitive job variables on re-run

### CODE REVIEW

File: .github/workflows/ci_pullrequest.yml [hunk 1]
```diff
@@ -1,250 +1,258 @@
...
-  pull_request:
-  workflow_call:
+# ---
+# ▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄
+# ██░▄▄▀██░▄▄▄░████░▀██░██░▄▄▄░█▄▄░▄▄████░▄▀▄░██░▄▄▄██░▄▄▀██░▄▄░██░▄▄▄
+# ██░██░██░███░████░█░█░██░███░███░██████░█░█░██░▄▄▄██░▀▀▄██░█▀▀██░▄▄▄
+# ██░▀▀░██░▀▀▀░████░██▄░██░▀▀▀░███░██████░███░██░▀▀▀██░██░██░▀▀▄██░▀▀▀
+# ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀
+#
+# This has CI paused to cut down on CI churn as we do a collaborative static code review.
```
Comment (bryanculver) [inline .github/workflows/ci_pullrequest.yml:8]: ☝️

File: nautobot/core/templates/graphene/graphiql.html [hunk 2]
```diff
@@ -30,6 +30,9 @@
   <link rel="stylesheet"
         href="{% static 'flatpickr-4.6.9/themes/light.min.css' %}"
         onerror="window.location='{% url 'media_failure' %}?filename=flatpickr-4.6.9/themes/light.min.css'">
+  <link rel="stylesheet" id="dark-theme"
+        href="{% static 'css/dark.css' %}?v{{ settings.VERSION }}"
+        onerror="window.location='{% url 'media_failure' %}?filename=css/dark.css'" disabled="disabled" />
   <link rel="stylesheet"
         href="{% static 'css/base.css' %}?v{{ settings.VERSION }}"
         onerror="window.location='{% url 'media_failure' %}?filename=css/base.css'">
```
Comment (glennmatthews) [inline nautobot/core/templates/graphene/graphiql.html:38]: Should replace this section with an `{% include 'inc/media.html' %}` similar to `nautobot/core/templates/base.html`.

File: nautobot/ipam/views.py [hunk 3]
```diff
@@ -316,6 +316,7 @@ class RoleListView(generic.ObjectListView):
         prefix_count=count_related(Prefix, "role"),
         vlan_count=count_related(VLAN, "role"),
     )
+    filterset = filters.RoleFilterSet
```
Comment (glennmatthews) [inline nautobot/ipam/views.py:319]: Nice to have but not required: add a new `RoleFilterForm`

File: nautobot/tenancy/views.py [hunk 4]
```diff
@@ -19,6 +19,7 @@ class TenantGroupListView(generic.ObjectListView):
     queryset = TenantGroup.objects.add_related_count(
         TenantGroup.objects.all(), Tenant, "group", "tenant_count", cumulative=True
     )
+    filterset = filters.TenantGroupFilterSet
```
Comment (glennmatthews) [inline nautobot/tenancy/views.py:22]: Nice to have but not required: add a `TenantGroupFilterForm`.

File: nautobot/utilities/testing/__init__.py [hunk 5]
```diff
@@ -114,6 +119,16 @@ def setUp(self):
...
+        # Create the test user and assign permissions
+        self.user = User.objects.create_user(username="testuser")
+        self.add_permissions(*self.user_permissions)
+
+        # Initialize the test client
+        self.client = Client()
+
+        # Force login explicitly with the first-available backend
+        self.client.force_login(self.user)
+
```
Comment (glennmatthews) [inline nautobot/utilities/testing/__init__.py:131]: Need to review why we're not just calling `super().setUpNautobot()` here - wasn't that the point of #1954?

File: pyproject.toml [hunk 6]
```diff
@@ -78,6 +78,8 @@ django-taggit = "~1.5.1"
 # Represent time zones in Django
 # NOTE: django-timezone-field 4.2.x is available but appears to break our initial migrations?
 django-timezone-field = "~4.1.2"
+# Tree database structures based on Common Table Expressions
+django-tree-queries = "~0.11"
```
Comment (glennmatthews) [inline pyproject.toml:82]: Are we OK with shipping 1.4 with it depending on *both* django-mptt and django-tree-queries?

Comment (bryanculver) [reply pyproject.toml:82]: IMHO: Yes. For now we have seen a bit of roadblocks with both (maintained vs feature compute) and we aren't providing/pushing those interfaces directly to users.

We still need to evaluate if we can leave `django-mptt` comfortably which I don't feel we need rush at the moment.

File: nautobot/docs/release-notes/version-1.4.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+#### Object Detail Tabs ([#1000](https://github.com/nautobot/nautobot/issues/1000))
+
+A plugin may now define extra tabs which will be appended to the object view's list of tabs.
+
+You can refer to the [plugin development guide](../plugins/development.md##adding-extra-tabs) on how to add tabs to existing object detail views.
+
+#### Improved Filter Coverage for DCIM models ([#1729](https://github.com/nautobot/nautobot/issues/1729))
+
+The DCIM FilterSets have been updated with 137 new filters, including hybrid filters that support filtering on both `pk` and `slug` (or `pk` and `name` where `slug` is not available). A new filter class `NaturalKeyOrPKMultipleChoiceFilter` was added to `nautobot.utilities.filters` to support filtering on multiple fields of a related object. See the [Best Practices](../development/best-practices/#mapping-model-fields-to-filters) documentation for more information.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-1.4.md:53]: Let's check the ordering of subheadings in the Release Overview section - in previous releases, for lack of any better order, we've listed major features in alphabetical order. Looks like we're deviating from that here.

File: pyproject.toml (hunk 6 above)
Comment (jathanism) [reply pyproject.toml:82]: No, but is it possible to rip out MPTT tho and still be backwards-compatible?

File: nautobot/docs/release-notes/version-1.4.md (hunk 7 above)
Comment (bryanculver) [reply nautobot/docs/release-notes/version-1.4.md:53]: Fixed in b34dbb3

File: nautobot/circuits/templates/circuits/circuit.html [hunk 8]
```diff
@@ -19,29 +19,23 @@
                 </tr>
                 <tr>
                     <td>Provider</td>
-                    <td>
-                        <a href="{% url 'circuits:provider' slug=object.provider.slug %}">{{ object.provider }}</a>
-                    </td>
+                    <td>{{ object.provider|hyperlinked_object }}</td>
```
Comment (gsnider2195) [inline nautobot/circuits/templates/circuits/circuit.html:22]: Do we need to add a change log entry for #1833? I'm not seeing one.

File: nautobot/docs/release-notes/version-1.4.md [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+#### Location Data Model ([#1052](https://github.com/nautobot/nautobot/issues/1052))
+
+To locate network information more precisely than a Site defines, you can now define a hierarchy of Location Types (for example, `Building` ← `Floor` ← `Room`) and then create Locations corresponding to these types within each Site. Data objects such as devices, prefixes, VLAN groups, etc. can thus be mapped or assigned to Location representing a specific building, wing, floor, room, etc. as appropriate to your needs.
+
+!!! info
+    At present, Locations fill the conceptual space between the more abstract Region and Site models and the more concrete Rack Group model. In a future Nautobot release, some or all of these other models may be collapsed into Locations. That is to say, in the future you might not deal with Regions and Sites as distinct models, but instead your Location Type hierarchy might include these higher-level categories, becoming something like Country ← City ← Site ← Building ← Floor ← Room.
+
+#### Object Detail Tabs ([#1000](https://github.com/nautobot/nautobot/issues/1000))
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:56]:
```suggestion
#### Custom Tabs in Object Detail View ([#1000](https://github.com/nautobot/nautobot/issues/1000))
```

Comment (bryanculver) [reply nautobot/docs/release-notes/version-1.4.md:56]: I agree with the suggestion and we should probably add a `TODO` to double check the alpha sort of the overview section before we go out.

File: nautobot/docs/release-notes/version-1.4.md [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+Nautobot's UI now supports dark mode, both explicitly and via browser preference selection.
+
+The "Theme" link in the footer provides a modal popup to select the preferred theme. This preference is saved per browser via `localStorage`.
+
+#### Improved Filter Coverage for DCIM, Virtualization Models
+
+- DCIM: [#1729](https://github.com/nautobot/nautobot/issues/1729)
+- Virtualization: [#1735](https://github.com/nautobot/nautobot/issues/1735)
+
+The DCIM, Virtualization FilterSets have been updated with over 150 new filters, including hybrid filters that support filtering on both `pk` and `slug` (or `pk` and `name` where `slug` is not available). A new filter class `NaturalKeyOrPKMultipleChoiceFilter` was added to `nautobot.utilities.filters` to support filtering on multiple fields of a related object. See the [Best Practices](../development/best-practices/#mapping-model-fields-to-filters) documentation for more information.
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:39]:
```suggestion
The DCIM and Virtualization FilterSets have been updated with over 150 new filters, including hybrid filters that support filtering on both `pk` and `slug` (or `pk` and `name` where `slug` is not available). A new filter class `NaturalKeyOrPKMultipleChoiceFilter` was added to `nautobot.utilities.filters` to support filtering on multiple fields of a related object. See the [Best Practices](../development/best-practices/#mapping-model-fields-to-filters) documentation for more information.
```

File: nautobot/docs/release-notes/version-1.4.md [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+!!! info
+    At present, Locations fill the conceptual space between the more abstract Region and Site models and the more concrete Rack Group model. In a future Nautobot release, some or all of these other models may be collapsed into Locations. That is to say, in the future you might not deal with Regions and Sites as distinct models, but instead your Location Type hierarchy might include these higher-level categories, becoming something like Country ← City ← Site ← Building ← Floor ← Room.
+
+#### Object Detail Tabs ([#1000](https://github.com/nautobot/nautobot/issues/1000))
+
+A plugin may now define extra tabs which will be appended to the object view's list of tabs.
+
+You can refer to the [plugin development guide](../plugins/development.md##adding-extra-tabs) on how to add tabs to existing object detail views.
+
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:61]: This feature should probably be highlighted in the release overview
```suggestion

#### Override Core and Plugin Views ([#1466](https://github.com/nautobot/nautobot/issues/1466))

Plugins can now override any of the core or plugin views by providing an `override_views` `dict` in a plugin's `views.py` file. For detailed information see the [plugin development guide](../plugins/development/#replacing-views)

```

File: nautobot/docs/release-notes/version-1.4.md [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+!!! info
+    At present, Locations fill the conceptual space between the more abstract Region and Site models and the more concrete Rack Group model. In a future Nautobot release, some or all of these other models may be collapsed into Locations. That is to say, in the future you might not deal with Regions and Sites as distinct models, but instead your Location Type hierarchy might include these higher-level categories, becoming something like Country ← City ← Site ← Building ← Floor ← Room.
+
+#### Object Detail Tabs ([#1000](https://github.com/nautobot/nautobot/issues/1000))
+
+A plugin may now define extra tabs which will be appended to the object view's list of tabs.
+
+You can refer to the [plugin development guide](../plugins/development.md##adding-extra-tabs) on how to add tabs to existing object detail views.
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:60]:
```suggestion
You can refer to the [plugin development guide](../plugins/development.md#adding-extra-tabs) on how to add tabs to existing object detail views.
```

File: nautobot/docs/release-notes/version-1.4.md (hunk 11 above)
Comment (bryanculver) [reply nautobot/docs/release-notes/version-1.4.md:61]: This isn't a hallmark feature that people should do very often. @lampwins Thoughts on including this in the release overview?

File: nautobot/docs/release-notes/version-1.4.md [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+- [#2137](https://github.com/nautobot/nautobot/issues/2137) - Fixed incorrect parameter name in `NaturalKeyOrPKMultipleChoiceFilter` documentation.
+- [#2142](https://github.com/nautobot/nautobot/pull/2142) - Fixed incorrect URL field in REST API nested relationship representation.
+- [#2165](https://github.com/nautobot/nautobot/pull/2165) - Fix up relationship-association API test issue.
+
+## v1.4.0b1 (2022-07-30)
+
+### Added
+
+- [#1463](https://github.com/nautobot/nautobot/issues/1463) - Added REST API support for opt-in `relationships` data on model endpoints; added `NautobotModelSerializer` base class.
+- [#1614](https://github.com/nautobot/nautobot/issues/1614) - Added support for nesting of Dynamic Groupsallowing inclusion/exclusion rules of sub-group members
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:162]:
```suggestion
- [#1614](https://github.com/nautobot/nautobot/issues/1614) - Added support for nesting of Dynamic Groups, allowing inclusion/exclusion rules of sub-group members
```

File: nautobot/docs/release-notes/version-1.4.md [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+- [#1917](https://github.com/nautobot/nautobot/pull/1917) - Update dependency svgwrite to ~1.4.2
+- [#1919](https://github.com/nautobot/nautobot/pull/1919) - Update dependency Pillow to ~9.1.1
+- [#1920](https://github.com/nautobot/nautobot/pull/1920) - Update dependency coverage to ~6.4.1
+- [#1921](https://github.com/nautobot/nautobot/pull/1921) - Update dependency django-auth-ldap to ~4.1.0
+- [#1924](https://github.com/nautobot/nautobot/pull/1924) - Update dependency django-cors-headers to ~3.13.0
+- [#1925](https://github.com/nautobot/nautobot/pull/1925) - Update dependency django-debug-toolbar to ~3.4.0
+- [#1928](https://github.com/nautobot/nautobot/pull/1928) - Update dependency napalm to ~3.4.1
+- [#1929](https://github.com/nautobot/nautobot/pull/1929) - Update dependency selenium to ~4.2.0
+- [#1945](https://github.com/nautobot/nautobot/issues/1945) - Change the `settings_and_registry` default context processor to purely `settings`, moving registry dictionary to be accessible via `registry` template tag.
+- [#2072](https://github.com/nautobot/nautobot/pull/2072) - Expand on `query_params` for `ObjectVar` in Jobs documentation.
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:225]: This was merged on 7/20 and should be moved to `1.4.0b1`

File: nautobot/docs/release-notes/version-1.4.md [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+### Added
+
+- [#1000](https://github.com/nautobot/nautobot/issues/1000) - Object detail views can now have extra UI tabs which are defined by a plugin.
+- [#1052](https://github.com/nautobot/nautobot/issues/1052) - Initial prototype implementation of Location data model
+- [#1318](https://github.com/nautobot/nautobot/issues/1318) - Added `nautobot.extras.forms.NautobotBulkEditForm` base class. All bulk-edit forms for models that support both custom fields and relationships now inherit from this class.
+- [#1466](https://github.com/nautobot/nautobot/issues/1466) - Plugins can now override views.
+- [#1729](https://github.com/nautobot/nautobot/issues/1729) - Add new filter class `NaturalKeyOrPKMultipleChoiceFilter` to `nautobot.utilities.filters`.
+- [#1729](https://github.com/nautobot/nautobot/issues/1729) - Add 137 new filters to `nautobot.dcim.filters` FilterSets.
+- [#1729](https://github.com/nautobot/nautobot/issues/1729) - Add `cable_terminations` to the `model_features` registry.
+- [#1875](https://github.com/nautobot/nautobot/issues/1875) - Add ability to quickly re-submit a previously run `Job` with the same parameters.
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:205]: This was merged on 7/21 and should be moved to `1.4.0b1`

File: nautobot/docs/release-notes/version-1.4.md [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+    Setting `STRICT_FILTERING` to `False` can result in unexpected filtering results in the case of user error, for example a request to `/api/dcim/devices/?has_primry_ip=false` (note the typo `primry`) will result in a list of all devices, rather than the intended list of only devices that lack a primary IP address. In the case of Jobs or external automation making use of such a filter, this could have wide-ranging consequences.
+
+#### Moved Registry Template Context ([#1945](https://github.com/nautobot/nautobot/issues/1945))
+
+The `settings_and_registry` default context processor was changed to purely `settings` - the (large) Nautobot application registry dictionary is no longer provided as part of the render context for all templates by default. Added a new `registry` template tag that can be invoked by specific templates to provide this variable where needed.
+
+## v1.4.0b2 (2022-MM-DD)
+
+### Added
+
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:140]:
```suggestion

- [#767](https://github.com/nautobot/nautobot/issues/767) - Added notes field to Primary and Organizational models
```

File: nautobot/docs/release-notes/version-1.4.md [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+- [#1735](https://github.com/nautobot/nautobot/issues/1735) - Added missing filters to model FilterSets for Virtualization models.
+- [#1865](https://github.com/nautobot/nautobot/issues/1865) - Added support for a custom template on Job forms.
+- [#1877](https://github.com/nautobot/nautobot/issues/1877) - Add new job base class JobHookReceiver to support triggering job execution from change events.
+- [#1883](https://github.com/nautobot/nautobot/issues/1883) - Add ability to filter objects by their relationships into the existing FilterSet.
+- [#1884](https://github.com/nautobot/nautobot/issues/1884) - Add ability to set the relationship filter via the filter form.
+- [#2035](https://github.com/nautobot/nautobot/pull/2035) - Added change source context to object change context manager.
+- [#2051](https://github.com/nautobot/nautobot/issues/2051) - Add changelog url for Relationships.
+- [#2061](https://github.com/nautobot/nautobot/pulls/2061) - Add draggable child groups to Dynamic Groups edit view in UI, recompute and hide weights.
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:170]:
```suggestion
- [#2061](https://github.com/nautobot/nautobot/pull/2061) - Add draggable child groups to Dynamic Groups edit view in UI, recompute and hide weights.
```

File: nautobot/docs/release-notes/version-1.4.md [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+- [#2103](https://github.com/nautobot/nautobot/pull/2103) - Add job hooks feature.
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:171]:
```suggestion
- [#1878](https://github.com/nautobot/nautobot/issues/1878) - Add job hooks feature.
```

File: nautobot/docs/release-notes/version-1.4.md [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+The "Theme" link in the footer provides a modal popup to select the preferred theme. This preference is saved per browser via `localStorage`.
+
+#### Improved Filter Coverage for DCIM, Virtualization Models
+
+- DCIM: [#1729](https://github.com/nautobot/nautobot/issues/1729)
+- Virtualization: [#1735](https://github.com/nautobot/nautobot/issues/1735)
+
+The DCIM, Virtualization FilterSets have been updated with over 150 new filters, including hybrid filters that support filtering on both `pk` and `slug` (or `pk` and `name` where `slug` is not available). A new filter class `NaturalKeyOrPKMultipleChoiceFilter` was added to `nautobot.utilities.filters` to support filtering on multiple fields of a related object. See the [Best Practices](../development/best-practices/#mapping-model-fields-to-filters) documentation for more information.
+
+#### Job Hooks ([#2103](https://github.com/nautobot/nautobot/pull/2103))
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:41]:
```suggestion
#### Job Hooks ([#1878](https://github.com/nautobot/nautobot/issues/1878))
```

File: nautobot/docs/release-notes/version-1.4.md (hunk 16 above)
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-1.4.md:140]:
```suggestion

- [#1498](https://github.com/nautobot/nautobot/issues/1498) - Added extended lookup expression filters to custom fields.
```

**Comment by christhant** [comment]
noticed some of the html tags don't have appropriate attribute values. such as alt text, label, ..etc. I would love to see if we use web accessibility checker such as Arc tool or scanner that can be incorporated with the ci/cd pipeline and give report on web accessibility compliant.

File: nautobot/circuits/models.py [hunk 20]
```diff
@@ -290,6 +295,13 @@ class CircuitTermination(PrimaryModel, PathEndpoint, CableTermination):
         blank=True,
         null=True,
     )
+    location = models.ForeignKey(
+        to="dcim.Location",
+        on_delete=models.PROTECT,
+        related_name="circuit_terminations",
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/circuits/models.py:301]: This is the kind of thing we have consistency problems all over the place. But all of the `related_name` for ForeignKeys on `CircuitTermination` are using `circuit_terminations` so this is at least self-consistent.

File: nautobot/core/templates/admin/base.html [hunk 21]
```diff
@@ -166,17 +169,26 @@ <h1>{% render_with_template_if_exist title|lower|add:"/admin_app_name.html" titl
...
                         <i class="mdi mdi-graphql text-primary"></i> <a href="{% url 'graphql' %}">GraphQL</a> &middot;
-                        <i class="mdi mdi-xml text-primary"></i> <a href="https://github.com/nautobot/nautobot">Code</a> &middot;
-                        <i class="mdi mdi-lifebuoy text-primary"></i> <a href="https://github.com/nautobot/nautobot/wiki">Help</a>
+                        <i class="mdi mdi-xml text-primary"></i> <a href="{{ settings.BRANDING_URLS.code }}">Code</a> &middot;
+                        <i class="mdi mdi-lifebuoy text-primary"></i> <a href="{{ settings.BRANDING_URLS.help }}">Help</a>
                     </p>
                 </div>
             </div>
         </div>
     </footer>
```
[the remark this answers is not in the document]
Comment (bryanculver) [reply nautobot/core/templates/admin/base.html:188]: Issue #2210

File: nautobot/core/templates/graphene/graphiql.html [hunk 22]
```diff
@@ -93,6 +96,8 @@
         onerror="window.location='{% url 'media_failure' %}?filename=flatpickr-4.6.9/flatpickr.min.js'"></script>
   <script src="{% static 'js/forms.js' %}?v{{ settings.VERSION }}"
         onerror="window.location='{% url 'media_failure' %}?filename=js/forms.js'"></script>
+  <script src="{% static 'js/theme.js' %}"
```
[the remark this answers is not in the document]
Comment (bryanculver) [reply nautobot/core/templates/graphene/graphiql.html:99]: Issue #2209

File: examples/example_plugin/example_plugin/templates/example_plugin/tab_device_detail_2.html [hunk 23]
```diff
@@ -0,0 +1,6 @@
+{% extends 'dcim/device.html' %}
```
[the remark this answers is not in the document]
Comment (bryanculver) [reply examples/example_plugin/example_plugin/templates/example_plugin/tab_device_detail_2.html:1]: Issue #2211

**Comment by bryanculver** [comment]
Feedback accounted for or migrated to new issues.

**Pull request closed without merging**